### PR TITLE
[APIV4] POST /emoji for apiV4

### PR DESCRIPTION
**Pull request by cpanato**
#### Summary
This PR implements the POST /emoji for apiV4 - upload custom emoji endpoint

#### Ticket Link
n/a

#### Checklist
- [x] Added or updated unit tests
- [x] Added API documentation - PR: https://github.com/mattermost/mattermost-api-reference/pull/187
- [x] All new/modified APIs include changes to the drivers

### CODE REVIEW

**Comment by cpanato** [comment]
investigating why sometimes it fails

File: app/emoji.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/mattermost/platform/model"
+)
+
+const (
+	MaxEmojiFileSize = 1000 * 1024 // 1 MB
+	MaxEmojiWidth    = 128
+	MaxEmojiHeight   = 128
+)
+
+func UploadEmojiImage(id string, imageData *multipart.FileHeader) *model.AppError {
```
Comment (jwilander) [inline app/emoji.go]: You refactored these out of the `api` package right? I know we're using these new functions with `api4` but can we use them with `api` as well? Just so we don't have duplicate code anywhere

File: api/emoji_test.go [hunk 2]
```diff
@@ -113,14 +110,14 @@ func TestCreateEmoji(t *testing.T) {
 	}
 
 	// try to create an emoji when they're disabled
-	if _, err := Client.CreateEmoji(emoji, createTestGif(t, 10, 10), "image.gif"); err == nil {
+	if _, err := Client.CreateEmoji(emoji, utils.CreateTestGif(t, 10, 10), "image.gif"); err == nil {
```
Comment (cpanato) [inline api/emoji_test.go]: @jwilander the apiv3 is using the refactor :)

File: app/emoji.go (hunk 1 above)
Comment (cpanato) [inline app/emoji.go]: ah the app, sure sure

Comment (cpanato) [inline app/emoji.go]: done

File: api4/emoji.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	props := m.Value
+
+	emoji := model.EmojiFromJson(strings.NewReader(props["emoji"][0]))
+	if emoji == nil {
+		c.SetInvalidParam("createEmoji")
+		return
+	}
+
+	// wipe the emoji id so that existing emojis can't get overwritten
+	emoji.Id = ""
```
Comment (crspeller) [inline api4/emoji.go]: I think some of this belongs in the app layer.

Comment (cpanato) [inline api4/emoji.go]: done

**Comment by grundleborg** [comment]
@cpanato please could you do a rebase to fix the conflicts?

@crspeller @jwilander the PR has been updated in response to your review comments.

**Comment by cpanato** [comment]
@grundleborg done

File: api4/emoji.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	newEmoji, err := app.CreateEmoji(c.Session.UserId, emoji, m)
+	if err != nil {
+		c.Err = err
+		return
+	} else {
+		w.Write([]byte(newEmoji.ToJson()))
+	}
+
+	// // wipe the emoji id so that existing emojis can't get overwritten
+	// emoji.Id = ""
```
Comment (crspeller) [inline api4/emoji.go]: I think you forgot to remove this!

Comment (cpanato) [inline api4/emoji.go]: sorry, i forgot that :(

Comment (cpanato) [inline api4/emoji.go]: done

File: app/emoji.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+	// do our best to validate the emoji before committing anything to the DB so that we don't have to clean up
+	// orphaned files left over when validation fails later on
+	emoji.PreSave()
+	if err := emoji.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if emoji.CreatorId != sessionUserId {
+		return nil, model.NewAppError("createEmoji", "api.emoji.create.other_user.app_error", nil, "", http.StatusUnauthorized)
```
Comment (jwilander) [inline app/emoji.go]: Can we change the status code here to StatusForbidden? StatusUnauthorized should be used when the user doesn't have an active session

File: app/emoji.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+		return nil, result.Err
+	} else {
+		return result.Data.(*model.Emoji), nil
+	}
+}
+
+func UploadEmojiImage(id string, imageData *multipart.FileHeader) *model.AppError {
+	file, err := imageData.Open()
+	if err != nil {
+		return model.NewLocAppError("uploadEmojiImage", "api.emoji.upload.open.app_error", nil, "")
```
Comment (jwilander) [inline app/emoji.go]: Can we use `NewAppError` here with a StatusBadRequest code? Same with all the below `NewLocAppError` uses?

File: app/emoji.go (hunk 5 above)
Comment (cpanato) [inline app/emoji.go]: done

File: app/emoji.go (hunk 6 above)
Comment (cpanato) [inline app/emoji.go]: done

**Pull request merged**